woocommerce/woocommerce-android
# Product creation: Add search functionality into category selection

Closes: https://github.com/woocommerce/woomobile-private/issues/340


### Description


This is a quick HACK Week project (1 day size) to add search to product category selection in product creation.

What it does:
1. Adds a search icon with `SearchView` to the product category selection screen,
2. Does API call to search when search is entered
3. Adds a "flat list" functionality where if the current category list has orphans (e.g: if search result returns only child categories), then display the whole list as a flat list instead of hierarchical.

---

### Technical discussion
**1. Implementation**
Currently, there are two implementations of category selection:
1. `ProductCategoriesFragment` used in product creation,
4. `ProductCategorySelectorFragment` used in coupon creation.

The latter one is more modern and has a component that already has a search functionality added. Initially I thought to implement its component to `ProductCategoriesFragment`, but I found the task to be non-trivial and definitely not a one-day HACK week work for me. One reason is that `ProductCategoriesFragment` is tightly coupled with the product creation process (e.g: each category selection automatically updates product draft without leaving the category screen).

So instead, I chose to replicate the search functionality from `ProductCategorySelectorFragment`, into `ProductCategoriesFragment`.

For future improvement, it'd be great to refactor `ProductCategoriesFragment` to use the component on `ProductCategorySelectorFragment`.

**2. Searching limitation**

The search is hardcoded to return maximum of 100 results, and doesn't support load more functionality. I think this is fine for a v1, and supporting more than 100 results can be considered an edge case.

---

### Testing instructions

1. Prepare a demo site with multiple categories and parent/sub-categories. It'd be good to have some parent and child categories that have similar names, for testing purposes.
2. Open app, then go to Products and select a Product,
3. Edit the product category,
4. Ensure search icon appears on top right,
5. Tap the icon, ensure that the search box appears,
6. Enter a search term, ensure that after a little bit, the search is being done and shows the result
7. Delete the search term, ensure that the list of categories are shown as before
8. Enter a search term where the result should be a mix of parent and child categories. Ensure the result is shown in a flat list.
9. Play around and check the selection state, it should remain consistent when searching and clearing search.


---

### Images/gif

[Screen_recording_20240318_220803.webm](https://github.com/woocommerce/woocommerce-android/assets/266376/ad6f6fa8-7cd4-47ab-ab86-32bcc457ca9c)
